### PR TITLE
Google clasp

### PR DESCRIPTION
**Pull request by michojel**
###### Motivation for this change

Upstream's [github page](https://github.com/google/clasp) says it all:

> 🗺️ Develop Locally: clasp allows you to develop your Apps Script projects locally. That means you can check-in your code into source control, collaborate with other developers, and use your favorite tools to develop Apps Script.
>
>🔢 Manage Deployment Versions: Create, update, and view your multiple deployments of your project.
>
> 📁 Structure Code: clasp automatically converts your flat project on script.google.com into folders.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by michojel** [comment]
Rebased.

Not sure whom to ping. @svanderburg would you be so kind to take a look or perhaps suggest another reviewer?
Note: this is my first contribution to NixOs.
Thanks!

**Comment by mmahut** [comment]
Any updates on this pull request, please?

**Comment by michojel** [comment]
Resurrected, but do not merge yet. Need to resolve a TypeError first:

```
clasp --version
/nix/store/q90cs1sqmjhg2sdksqal5qpyryag9hhs-node-_at_google_slash_clasp-2.2.1/lib/node_modules/@google/clasp/src/dotfile.js:97
    RC: dotf(exports.DOT.RC.DIR, exports.DOT.RC.NAME),
        ^

TypeError: dotf is not a function
    at Object.<anonymous> (/nix/store/q90cs1sqmjhg2sdksqal5qpyryag9hhs-node-_at_google_slash_clasp-2.2.1/lib/node_modules/@google/clasp/src/dotfile.js:97:9)
    at Module._compile (internal/modules/cjs/loader.js:701:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:712:10)
    at Module.load (internal/modules/cjs/loader.js:600:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:539:12)
    at Function.Module._load (internal/modules/cjs/loader.js:531:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at Object.<anonymous> (/nix/store/q90cs1sqmjhg2sdksqal5qpyryag9hhs-node-_at_google_slash_clasp-2.2.1/lib/node_modules/@google/clasp/src/utils.js:52:17)
    at Module._compile (internal/modules/cjs/loader.js:701:30)
```

**Comment by michojel** [comment]
> Resurrected, but do not merge yet. Need to resolve a TypeError first:
> 
Fixed by a patch.

@turboMaCk, @domenkozar PTAL (adding people having some experience with node packages, feel free to tag others)

**Review by turboMaCk — APPROVED** [review]
I'm not a commiter btw so just a review.

- [x] builds and runs on NixOS
- [x] builds and runs on MacOS

I've found 2 minor things in the diff I'm unsure about but that's for someone more experienced to decide. I don't think either is a blocker for merge.

I would recommend squishing commits before merge.

**Comment by turboMaCk** [comment]
@GrahamcOfBorg build google-clasp

File: pkgs/development/misc/google-clasp/default.nix
Comment (michojel) [reply pkgs/development/misc/google-clasp/default.nix]: Thanks @turboMaCk. The unpacking of node packages takes place in the `installPhase`:

```
checking outputs of '/nix/store/bd9fs5z96ypszcn0zqmh9lp2l29m51sp-node-_at_google_slash_clasp-2.2.1.drv'...                                                                  
unpacking sources                                                                                                                                                            
patching sources                                                                                                                                                             
configuring                                                                                                                                                                 
no configure script, doing nothing                                                                                                                                           
building                                                                                                                                                                     
installing                                                                                                                                                                   
unpacking source archive /nix/store/xzbw3n2j4zq5nh1j4dkx0aqn1mr629lp-clasp-2.2.1.tgz                                                                                         
unpacking source archive /nix/store/ibwdid42ly8x7xidcvp0jr7jnnm281lf-fs.scandir-2.1.1.tgz                                                                                    
unpacking source archive /nix/store/zfpb0hszl9827qmhbkhy3sksrlyfnja0-fs.stat-2.0.1.tgz                                                                                       
unpacking source archive /nix/store/33qz58lq0f86hcbdi13jp04ndwp4c824-fs.walk-1.2.2.tgz                                                                               
...
```

The `patchPhase` does not have the sources available yet.

The nodePackages seem to patch the sources during in the `preBuild` phase:

- https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/node-packages/default-v10.nix#L47
- https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/node-packages/default-v10.nix#L71

So this can be inconsistent either with a regular nix package build or node-package build. An alternative approach to this would be:
1. list the package in [node-packages-v10.json ](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/node-packages/node-packages-v10.json)
2. add another override like ([pnpm = nodePackages.pnpm.override](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/node-packages/default-v10.nix#L67)) in the [default-v10.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/node-packages/default-v10.nix)

I don't have a strong opinion on this one.

@adisbladis toughts?

File: pkgs/development/misc/google-clasp/generate.sh
Comment (michojel) [reply pkgs/development/misc/google-clasp/generate.sh]: Two other occurrences:
- https://github.com/NixOS/nixpkgs/blob/0e296d5fcdaf32dd113e49738dfba6216c01a8da/pkgs/tools/package-management/nixui/generate.sh#L4
- https://github.com/NixOS/nixpkgs/blob/0e296d5fcdaf32dd113e49738dfba6216c01a8da/pkgs/tools/networking/airfield/deps.sh#L6

From my point of view, the less redundancy, the better.

**Comment by michojel** [comment]
@Ma27 could you please take a look?

File: pkgs/development/misc/google-clasp/default.nix
Comment (turboMaCk) [reply pkgs/development/misc/google-clasp/default.nix]: I was playing with this and I think the solution you come up with is fine and most straight forward at the moment.

File: pkgs/development/misc/google-clasp/generate.sh
Comment (turboMaCk) [reply pkgs/development/misc/google-clasp/generate.sh]: thanks for references. My thinking was that this creates non obvious coupling between what appears as unrelated modules. Anyway this is not really up to me to decide so unless someone more experienced comes in and gives definite answer I would keep it as is.

File: pkgs/development/misc/google-clasp/default.nix
Comment (Ma27) [reply pkgs/development/misc/google-clasp/default.nix]: Seems fine for now. However I'd like to refresh my knowledge about the node2nix codebase first to check if there's a proper workaround (otherwise I'd file an issue there later).

File: pkgs/development/misc/google-clasp/generate.sh
Comment (Ma27) [reply pkgs/development/misc/google-clasp/generate.sh]: IIRC we use a single node2nix in nixpkgs (hence the same node-env.nix) for everything, so this should be fine IMHO.
In the end we already have a lot of large diffs because of that, so I'd prefer to use the node-env.nix we already have **unless** we encounter issues that can be resolved by committing another one : )

**Comment by Ma27** [comment]
After skimming through the diff, this appears to be fine. Before merging, I'd love to test this first. I'll probably be able to do this tonight, as soon as I have a better uplink to download all needed dependencies :)

**Review by Ma27 — APPROVED** [review]
LGTM. Unless there are further objections, I'd merge tomorrow :)

**Comment by michojel** [comment]
Thanks @Ma27 and @turboMaCk!

**Pull request merged**
